### ISSUE TITLE
conv_transpose_mkldnn has accuracy diff

### ISSUE THREAD

**Issue by luotao1** [PaddlePaddle/Paddle#17535]
`conv_transpose_mkldnn_op` has accuracy diff on our internal face model.
commit: 3fe2def is OK.
commit: 1eb1517 is wrong.
If I remove `conv_transpose` in op_lists, the accuracy is OK.
```
if (use_mkldnn) {
     cfg.EnableMKLDNN();
     std::unordered_set<std::string> op_list={"conv", "batch_norm", "relu", "leaky_relu", "elementwise_add", "tanh"};
     cfg.SetMKLDNNOp(op_list);
   }
```

**Comment by jczaja** [comment]
@luotao1 I understand that exact commit that introduce inaccuracy is  unknown. Since we do not have access to your face workload, is it possible your validation narrow down and trace commit that introduce a problem? Or perhaps some public test can be run to show inaccuracy?

**Comment by jczaja** [comment]
I can see that conv2d transpose of MKL-DNN was merged 29th of November 2018. So this suggest that inaccuracy
was there from the moment of introduction of conv2d transpose mkl-dnn.

**Comment by luotao1** [comment]
I test accuracy when your PR merged https://github.com/PaddlePaddle/Paddle/pull/14608#issuecomment-442442212, and it runs OK on accuracy diff.
I will see how to give you more clues about this issue.
By the way, do you have any suggestion?

**Comment by luotao1** [comment]
I replace the `conv_transpose_mkldnn_op` with https://github.com/jczaja/Paddle/blob/1540df51cf5b88a08afec4e94d5fc86d4ddfe5c5/paddle/fluid/operators/conv_transpose_mkldnn_op.cc from #14608, but it still has accuracy diff.

**Comment by jczaja** [comment]
My understanding:
1. So when you roll back to conv transpose 2d mkldnn as it was introduced all is fine. 
2. When you put conv2d transpose mkldnn from PR as it was merged into current develop there is a problem.

If this is what is seems then:
a) Meantime MKL-DNN was updated and then have implemented efficient JIT version for conv2d transpose , perhaps there is mistake there. Moving back mkl-dnn is not an option as INT8 kerenls were added. I can prepare (tomorow when I'm in a office)
branch with disabled efficent deconvolution (disabling inside mkl-dnn) and we can see if accuracy diff is fine.

Meantime:
1) please provide MKLDNN_VERBOSE=1 log from develop
So I can see which implementation is called on develop where problems are reported.

**Comment by ghost** [comment]
@luotao1 It's suspected that JIT version deconv MKLDNN kernel may have issue which cause this diff since no diff when using reference deconv MKLDNN kernel. @jczaja is co-working with MKLDNN team on this. He may ask for more MKLDNN verbose log soon.

**Comment by jczaja** [comment]
Modified MKL-DNN (no jitted conv transpose) when used on face model produce no error.
So two options are:
1) MKL-DNN jitted deconv (conv transpose) does not work properly for our use case. MKL-DNN tool
benchDNN was used to reproduce problem, but no reproduction so far (BenchDNN is using relative metric to test accuracy while paddle is using absolute value metric).
2) paddlepaddle conv transpose mkldnn op does not work for jitted formats. This is not very likely as the same code is shared with convolution mkldnn op and it works fine.


Next Steps:
1) Investigate which of conv transpose ops in face model is introducing most inaccuracy &
 Dump inputs and weights for that op & modify mkl-dnn benchDnn with those data

**Comment by jczaja** [comment]
It seems that conv2d_transpose mkl-dnn is working fine. Problem is elementwise_add mkldnn op.
To verify that one can run Face model with:
```
 std::unordered_set<std::string> op_list={"conv2d_transpose"};
 cfg.SetMKLDNNOp(op_list);

```
And this work fine. But when We run with two ops:
```
 std::unordered_set<std::string> op_list={"conv2d_transpose", "elementwise_add"};
 cfg.SetMKLDNNOp(op_list);

```
Then problem appears. Previous observation that MKL-DNN conv_transpose Jitted is broken as oppose to reference MKL-DNN conv_transpose was not precise. Jitted conv_transpose seems to be using diffrent mkl-dnn format (data arrangements) which is not handled correctly by elementwise_add mkl-dnn. 

To fix that we have two options:
1) Investigate and fix elementwise_add mkldnn
2) Make a pass merging conv2d_transpose and elementwise_add eg conv_transpose_bias_mkldnn_pass. 

I will work on option 2. As it should be better for face model performance and this task hopefully will takes less time to execute that further investigation of elementwise_add mkldnn and fixing it.

**Comment by jczaja** [comment]
While looking at Face model I noticed that some improvements to mkl-dnn integration in Paddle can be made to further speed up the model's execution:
1) Add leaky-relu mkl-dnn support: #17654 
2) Extend mkl-dnn conv passes to work with conv2d_transpose: #17655

**Comment by luotao1** [comment]
Thanks very much for your investigation and solutions!
For "Problem is elementwise_add mkldnn op", is #16880 solved this?

**Comment by jczaja** [comment]
@luotao1 #16880 Does not fix inaccuracy in Face model

**Comment by jczaja** [comment]
Solution to problem by making pass combining conv transpose and elementwise_add is almost done.

@luotao1 Please check code from PR #17644 if it works for this issue and other internal workloads.

**Comment by ghost** [comment]
@jczaja Thanks for your effort on PR #17644. Could you explain a little bit more about why a conv_transpose + elementwise_add (withOUT fusion) cause this issue?

**Comment by luotao1** [comment]
> Please check code from PR #17644 if it works for this issue and other internal workloads

@jczaja Thanks very much! I test this PR on all internal workloads, they run OK.

**Comment by ghost** [comment]
Fixed by #17644
